develop/items/custom-data-components.md

```diff
@@ -0,0 +1,267 @@
+);
+```
+
+You can see that we're now passing `<Integer>` as our generic type, indicating that this component will be stored as a single `int` value. For our codec, we are using the provided `Codec.INT` codec. We can get away with using basic codecs for simple components like this but more complex scenarios might require a custom codec (this will be covered briefly later on).
+
+::: info
+This example uses the `LightningStick` class created in the [Custom Item Interactions](./custom-item-interactions) page.
+:::
+
+If you start the game you should be able to enter a command like so:
+
+![Give command showing custom component](/assets/develop/items/custom_component_0.png)
+
+When you run the command you should recieve the item containing the component. However we are not currently using our component to do anything useful. Let's start by reading the value of the component in a way we can see.
+
+## Reading component value {#reading-component-value}
+
+Let's extend the `LightningStick` tooltip code to display the current value of the click count when we hover over our item in the inventory. We can use the `get()` method on our `ItemStack` to get our component value like so:
+
+```java
+int clickCount = stack.get(ModItems.CLICK_COUNT_COMPONENT);
+```
+
+This will return the current component value as the type which we defined when we registered our component. We can then use this value to add a tooltip entry. Add this line to the `appendTooltip` method in the item class:
+
+```java
+@Override
+public void appendTooltip(ItemStack stack, TooltipContext context, List<Text> tooltip, TooltipType type) {
+    int clickCount = stack.get(ModItems.CLICK_COUNT_COMPONENT);
+    tooltip.add(Text.translatable("item.modid.lightning_stick.used", clickCount).formatted(Formatting.GOLD));
+}
+```
+
+Don't forget to update your lang file (`/assets/<mod id>/lang/en_us.json`) and add this line:
+
+```json
+{
+    "item.modid.lightning_stick.used": "Used %1$s times"
+}
+```
+
+If you start the game and hover over the item you gave yourself earlier, you should see the count displayed as a tooltip!
+
+![Tooltip showing "Used 5 times"](/assets/develop/items/custom_component_1.png)
+
+However, if you give yourself a new Lightning Stick item *without* the custom component, the game will crash when you hover over the item in your inventory. You should see an error like this in the crash report:
+
+```
+java.lang.NullPointerException: Cannot invoke "java.lang.Integer.intValue()" because the return value of "net.minecraft.item.ItemStack.get(net.minecraft.component.ComponentType)" is null
+        at com.example.LightningStick.appendTooltip(LightningStick.java:45)
+        at net.minecraft.item.ItemStack.getTooltip(ItemStack.java:767)
+```
+
+As expected, since the `ItemStack` doesn't currently contain an instance of our custom component, calling `stack.get()` with our component type will return `null`. 
+
+There are two solutions we can use to address this problem:
```

Comment:
There's another solution: properly handling the absense of the component (by hiding the tooltip)

Reply:
Sure, I'll suggest it as an alternative